src/validation.cpp

```diff
@@ -1288,6 +1300,31 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
                    [](const auto& tx) { return Workspace(tx); });
     std::map<uint256, MempoolAcceptResult> results;
 
+    if (const auto v3_violation{CheckV3Inheritance(txns)}) {
```

Comment:
Not sure if I'm missing something, but this only checks that the v3 inheritance rules are enforced among transactions in the package.  What if someone sent a package of 2 v3 transactions, but the child transaction also spent a transaction that was already in the mempool?  I think the `CheckV3Inheritance()` needs to be able to pull parents from both the mempool and from the package, perhaps?

Reply:
the other variant you're referring to here https://github.com/bitcoin/bitcoin/pull/28948#discussion_r1421614259 should cover that case.

Reply:
Looks like it's broken, see this test:
```
diff --git a/test/functional/mempool_accept_v3.py b/test/functional/mempool_accept_v3.py
index 769b177cf523..713a8cb7406b 100755
--- a/test/functional/mempool_accept_v3.py
+++ b/test/functional/mempool_accept_v3.py
@@ -229,11 +229,42 @@ class MempoolAcceptV3(BitcoinTestFramework):
         tx_replacer = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=utxo_confirmed, fee_rate=DEFAULT_FEE * 10)
         self.check_mempool([tx_replacer["txid"]])
 
+    @cleanup
+    def test_package_v3_inheritance(self):
+        """
+        A v3 transaction in a package cannot have 2 v3 parents.
+        Test that if we have a transaction graph A -> B -> C, where A, B, C are
+        all v3 transactions, that we cannot use submitpackage to get the
+        transactions all into the mempool.
+
+        Verify, in particular, that if A is already in the mempool, then
+        submitpackage(B, C) will fail.
+        """
+        node = self.nodes[0]
+        self.log.info("Test that submitpackage won't allow v3 topology rules to be violated")
+        # This is our transaction "A":
+        tx_in_mempool = self.wallet.send_self_transfer(from_node=node, version=3)
+
+        # Verify that A is in the mempool
+        self.check_mempool([tx_in_mempool["txid"]])
+
+        # tx_0fee_parent is our transaction "B"; just create it.
+        tx_0fee_parent = self.wallet.create_self_transfer(utxo_to_spend=tx_in_mempool["new_utxo"], fee=0, fee_rate=0, version=3)
+
+        # tx_child_violator is our transaction "C"; create it:
+        tx_child_violator = self.wallet.create_self_transfer_multi(utxos_to_spend=[tx_0fee_parent["new_utxo"]], version=3)
+
+        # submitpackage(B, C) should fail
+        result = node.submitpackage([tx_0fee_parent["hex"], tx_child_violator["hex"]])
+        assert(result['package_msg'] != 'success')
+        #self.check_mempool([tx_in_mempool["txid"], tx_0fee_parent["txid"], tx_child_violator["txid"]])
+
     def run_test(self):
         self.log.info("Generate blocks to create UTXOs")
         node = self.nodes[0]
         self.wallet = MiniWallet(node)
         self.generate(self.wallet, 110)
+        self.test_package_v3_inheritance()
         self.test_v3_acceptance()
         self.test_v3_replacement()
         self.test_v3_bip125()
```

Reply:
that's the known topology issue IIUC: https://github.com/bitcoin/bitcoin/pull/28948#discussion_r1409842026

Reply:
Ah,  I see.  I think this is a different set of checks that fail, but I'm not sure exactly what the fix is so maybe both of these can be fixed at the same time...

Reply:
Added a check for in-package + mempool ancestors, and both of your tests.